File: Jump_Game_Greedy.py
# Greedy Approach
# Time complexity : O(n)
# Space complexity : O(1)
# Leetcode : Solved and submitted

class Solution:
    def canJump(self, nums: List[int]) -> bool:
      
        # find the length of the array and make target as the last index
        n = len(nums)
        target = n-1
        
        # start from the last index until we reach the first index
        for i in range(n-2,-1,-1):
          
            # add the current index with the value at current index and check if it is greater than the target, if so then make the new target as i
            if i + nums[i] >= target:
                target = i
        
        # at the end, if target is 0, then we can reach the end from the start, so return True else False
        if target == 0:
            return True
        return False

File: JumpGame_Greedy.java
//Time Complexity: O(N)
//Space Complexity : O(1)
//Method: Greedy approach
//come from the last , check if last element can be reached by any of the elements to the left, lly for all the elements to the left which implies, currIdx + nums[currIdx] >= nums.length -1, if this expression is valid, then the new target would be currIdx in the place of nums.length-1; target should be equal to 0 for jump possibility.

class Solution {
    public boolean canJump(int[] nums) {
        //base
        if(nums.length < 2)return true;
        int target = nums.length -1;

        for(int currIdx = nums.length -2 ; currIdx >= 0 ; currIdx--){
            if(currIdx + nums[currIdx] >= target){
                target = currIdx;
            }
            if(target==0)return true;
        }


        return false;
    }
}